[CGPO] Mixture of judges

gaetanlop opened the pull request:
# What does this PR do?

This PR adds the Mixture of judges part of the CGPO paper (https://arxiv.org/pdf/2409.20370). The judges are described in section 4.1.4 and the mixture of judges simply labels a generation as “violated” (0) if it fails any one of the constraint judgments and “satisfied” (1) otherwise.

Related to #2156 

## Before submitting
- [X] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [X] Did you write any new necessary tests?

## Who can review?

@kashif @lewtun

gaetanlop commented:
Still a draft PR

qgallouedec commented on trl/trainer/judges.py:
```diff
@@ -134,6 +134,46 @@ def judge(self, prompts: List[str], completions: List[List[str]], shuffle_order:
+    """
+
+    @abstractmethod
+    def judge(self, prompts: List[str], completions: List[str], shuffle_order: bool = True) -> List[int]:
```
Why not :?
```suggestion
    def judge(self, prompts: List[str], completions: List[str], shuffle_order: bool = True) -> List[bool]:
```

gaetanlop replied:
The judge could also return -1 to indicate that the inner process used to compute the preference has failed. However, the MixtureOfJudges should return a boolean. I will fix it.

qgallouedec commented on trl/trainer/judges.py:
```diff
@@ -134,6 +134,46 @@ def judge(self, prompts: List[str], completions: List[List[str]], shuffle_order:
         raise NotImplementedError("Judge subclasses must implement the `judge` method.")
 
 
+class BaseConstraintJudge(BaseJudge):
```
I'm trying to think wider than the paper. What could be a better name for such a judge? (ie a judge that returns 0/1)? `BinaryJudge`?

kashif replied:
yes in the nash-md terminology its a "hard-judge"

qgallouedec replied:
soft-judge is when the judge returns a value in [0, 1]?

gaetanlop replied:
> I'm trying to think wider than the paper. What could be a better name for such a judge? (ie a judge that returns 0/1)? `BinaryJudge`?

Yeah, that's how I named it in the first commits.

qgallouedec commented on trl/trainer/judges.py:
```diff
@@ -303,3 +343,27 @@ def get_rank(prompt, candidates):
         return ranks
+
+
+class MixtureOfConstraintJudges(BaseConstraintJudge):
```
Same question here: `MixtureConjunctionJudge`? `ConjunctionJudge`? `AllTrueJudge`?

gaetanlop replied:
`AllTrueJudge` seems cool in my opinion

qgallouedec commented on trl/trainer/judges.py:
```diff
@@ -303,3 +343,27 @@ def get_rank(prompt, candidates):
+
+    This class returns 0 ("violated") if it fails on any of the constraint judges (ie a judge returns 0 or -1) and returns 1 ("satisfied") otherwise.
+
+    It is an implementation of the Mixture of Judges as described in the CGPO paper: https://arxiv.org/pdf/2409.20370
```
```suggestion
    It is an implementation of the Mixture of Judges as described in the [CGPO paper](https://huggingface.co/papers/2409.20370)
```

qgallouedec commented on trl/trainer/judges.py:
```diff
@@ -303,3 +343,27 @@ def get_rank(prompt, candidates):
+    It is an implementation of the Mixture of Judges as described in the CGPO paper: https://arxiv.org/pdf/2409.20370
+
+    Args:
+        judges (List[BaseConstraintJudge]): A list of BaseConstraintJudge.
```
```suggestion
        judges (`List[BaseConstraintJudge]`): A list of [`BaseConstraintJudge`].
```

qgallouedec commented:
Thanks a lot @gaetanlop. Added some suggestion and open questions

gaetanlop commented:
Hey @qgallouedec, thanks for the review. I have added a `gold_answers` parameter to the judge function for both the `base` and `MoJs` class. Also, I have added a `safety_judge` and `factuality_judge` as described in the CGPO paper. They also have rule based judges, but in my opinion they are a little bit too tailored to specific tasks (coding/maths) to be added to the `trl` library. If you think the `factuality` and `safety` judges are also too specific to be in the lib I can remove them from the PR.

For the naming of the judges, let's do `BaseBinaryJudge` for the base class and `AllTrueJudge` for the moj following your suggestions?

qgallouedec commented on trl/trainer/judges.py:
```diff
@@ -303,3 +476,31 @@ def get_rank(prompt, candidates):
+        self.judges = judges
+
+    def judge(
+        self, prompts: List[str], completions: List[str], gold_answers: List[str] = None, shuffle_order: bool = True
```
can the gold_answer be none? If not I'd make it required arg.

gaetanlop replied:
Yes, the SafetyJudge of the paper do not use a gold answer for example

qgallouedec commented:
> Hey @qgallouedec, thanks for the review. I have added a `gold_answers` parameter to the judge function for both the `base` and `MoJs` class. Also, I have added a `safety_judge` and `factuality_judge` as described in the CGPO paper. They also have rule based judges, but in my opinion they are a little bit too tailored to specific tasks (coding/maths) to be added to the `trl` library. If you think the `factuality` and `safety` judges are also too specific to be in the lib I can remove them from the PR.

Thanks a lot! Nice work!

> For the naming of the judges, let's do `BaseBinaryJudge` for the base class and `AllTrueJudge` for the moj following your suggestions?

LGTM.

---

WDYT of having generic classes in `trl.judges` (`AllTrueJudge`, `BinaryJudge` etc.) and subclass them in `trl.trainer.cgpo_trainer` to get `SafetyConstraintJudge` and `FacultyConstraintJudge`? If in the future we need these classes elsewhere, we can still move them in `trl.judges`

gaetanlop commented:
Ok, I have removed the `FacultyConstraintJudge`  and the `SafetyConstraintJudge` from the PR and made the required renamings. Thanks @qgallouedec for the feedback.

qgallouedec commented on trl/trainer/judges.py:
```diff
@@ -135,6 +135,49 @@ def judge(self, prompts: List[str], completions: List[List[str]], shuffle_order:
+
+    @abstractmethod
+    def judge(
+        self, prompts: List[str], completions: List[str], gold_answers: List[str] = None, shuffle_order: bool = True
```
```suggestion
        self, prompts: List[str], completions: List[str], gold_answers: Optional[List[str]] = None, shuffle_order: bool = True
```

qgallouedec commented on trl/trainer/judges.py:
```diff
@@ -135,6 +135,49 @@ def judge(self, prompts: List[str], completions: List[List[str]], shuffle_order:
+
+    @abstractmethod
+    def judge(
+        self, prompts: List[str], completions: List[str], gold_answers: List[str] = None, shuffle_order: bool = True
```
maybe `"gold_completion"` or even `"ref_completion"` would be more suited, what do you think?

gaetanlop replied:
`"gold_completions"` seems good to me. I have done the update.

qgallouedec commented on trl/trainer/judges.py:
```diff
@@ -135,6 +135,53 @@ def judge(self, prompts: List[str], completions: List[List[str]], shuffle_order:
+        """
+        Judge the completion for a given prompt. Used to assess if a completion satisfies a constraint.
+
+        This base class should be used to implement binary evaluations as done in section 4.1.4 of the CGPO paper (https://arxiv.org/pdf/2409.20370).
```
```suggestion
        This base class should be used to implement binary evaluations as done in section 4.1.4 of the
        [CGPO paper](https://huggingface.co/papers/2409.20370).
```

qgallouedec commented on trl/trainer/judges.py:
```diff
@@ -361,3 +408,35 @@ def get_rank(prompt, candidates):
+        completions: List[str],
+        gold_completions: Optional[List[str]] = None,
+        shuffle_order: bool = True,
+    ) -> List[bool]:
```
This should return a list of int to be consistent with the super class

qgallouedec commented on trl/trainer/judges.py:
```diff
@@ -361,3 +408,35 @@ def get_rank(prompt, candidates):
+
+
+class AllTrueJudge(BaseBinaryJudge):
+    """
+    Unify the decision of multiple BaseBinaryJudge.
+
+    This class returns False if it fails on any of the binary judges (ie a judge returns 0 or -1) and returns True otherwise.
+
+    It is an implementation of the Mixture of Judges as described in the [CGPO paper](https://huggingface.co/papers/2409.20370)
+
+    Args:
+        judges (`List[BaseBinaryJudge]`): A list of [`BaseBinaryJudge`].
+    """
```
```suggestion
     """
     Unify the decision of multiple [`BaseBinaryJudge`] instances.

     Returns `1` only if all inner binary judges return `1`. If any judge returns `0`, it returns `0`.  
     If any judge returns `-1`, indicating a failure in its process, this judge will also return `-1`.

     Implements the Mixture of Judges as described in the [CGPO paper](https://huggingface.co/papers/2409.20370).

     Args:
     judges (`List[BaseBinaryJudge]`): A list of [`BaseBinaryJudge`] instances whose decisions will be unified.
     """
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/trl/pr_2159). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

qgallouedec commented:
LGTM, thanks @gaetanlop

qgallouedec commented:
We're having the issue

```
FAILED tests/test_nash_md_trainer.py::TestNashMDTrainer::test_nash_md_trainer_judge_training_1_conversational_prompt_only - ValueError: Cannot find pytorch_model.bin or model.safetensors in C:\Users\runneradmin\.cache\huggingface\hub\llm-blender\PairRM
```

that is mainly due to the fact that PairRM is requested for download simultaneously for different tests. This happens quite randomly, and is not related to any actual bug in the code base. I'll ignore it and merge.
